DataFormats/Portable/interface/PortableHostCollection.h

```diff
@@ -0,0 +1,81 @@
+#ifndef DataFormats_Portable_interface_PortableHostCollection_h
+#define DataFormats_Portable_interface_PortableHostCollection_h
+
+#include <optional>
+
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/host.h"
+
+// generic SoA-based product in host memory
+template <typename T>
+class PortableHostCollection {
+public:
+  using Layout = T;
+  using View = typename Layout::View;
+  using ConstView = typename Layout::ConstView;
+  using Buffer = alpaka::Buf<alpaka_common::DevHost, std::byte, alpaka::DimInt<1u>, uint32_t>;
+  using ConstBuffer = alpaka::ViewConst<Buffer>;
+
+  PortableHostCollection() = default;
+
+  PortableHostCollection(int32_t elements, alpaka_common::DevHost const &host)
+      // allocate pageable host memory
+      : buffer_{alpaka::allocBuf<std::byte, uint32_t>(
+            host, alpaka::Vec<alpaka::DimInt<1u>, uint32_t>{Layout::computeDataSize(elements)})},
+        layout_{buffer_->data(), elements},
+        view_{layout_} {
+    // Alpaka set to a default alignment of 128 bytes defining ALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128
+    assert(reinterpret_cast<uintptr_t>(buffer_->data()) % Layout::alignment == 0);
+  }
+
+  template <typename TDev>
+  PortableHostCollection(int32_t elements, alpaka_common::DevHost const &host, TDev const &device)
+      // allocate pinned host memory, accessible by the given device
+      : buffer_{alpaka::allocMappedBuf<std::byte, uint32_t>(
+            host, device, alpaka::Vec<alpaka::DimInt<1u>, uint32_t>{Layout::computeDataSize(elements)})},
+        layout_{buffer_->data(), elements},
+        view_{layout_} {
+    // Alpaka set to a default alignment of 128 bytes defining ALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128
+    assert(reinterpret_cast<uintptr_t>(buffer_->data()) % Layout::alignment == 0);
+  }
+
+  ~PortableHostCollection() = default;
+
+  // non-copyable
+  PortableHostCollection(PortableHostCollection const &) = delete;
+  PortableHostCollection &operator=(PortableHostCollection const &) = delete;
+
+  // movable
+  PortableHostCollection(PortableHostCollection &&other) = default;
+  PortableHostCollection &operator=(PortableHostCollection &&other) = default;
+
+  View &operator*() { return view_; }
+
+  View const &operator*() const { return view_; }
+
+  View *operator->() { return &view_; }
+
+  View const *operator->() const { return &view_; }
```

Comment:
If `operator*()` is changed to return `ConstView`, I'm not really sure what to suggest for this method. If `View` is copyable, as-is this method can give non-`const` access to data of a `PortableHostCollection const`.

Reply:
The implementation that I did not commit return `std::optional<ConstView>` in order to have something that supports `opertator*`...

Yesterday I think I had a better idea: make `View` inherit from `ConstView` and add the mutable accessors.
This way, a `View&` or `View*` can automatically be converted to `ConstView&`/`ConstView const&` or `ConstView*`/`ConstView const*`.

A proof-of-principle test is here: https://godbolt.org/z/E3jje3M5b .